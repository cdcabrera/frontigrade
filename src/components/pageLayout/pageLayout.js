import React from 'react';
import PropTypes from 'prop-types';
import {
  BackgroundImage,
  Brand,
  Page,
  PageHeader,
  PageSection,
  Toolbar,
  ToolbarGroup,
  ToolbarItem,
  Dropdown,
  DropdownToggle,
  DropdownItem,
  KebabToggle
} from '@patternfly/react-core';
import { HelpIcon, UserIcon } from '@patternfly/react-icons';
import { connect, reduxActions, reduxTypes, store } from '../../redux';
import helpers from '../../common/helpers';
import titleImg from '../../styles/images/title.svg';
import titleImgBrand from '../../styles/images/title-brand.svg';
import bgImages from '../../constants/patternflyConstants';

class PageLayout extends React.Component {
  state = {
    isMobile: false,
    isHelpDropdownOpen: false,
    isUserDropdownOpen: false,
    isMobileDropdownOpen: false
  };

  onAbout = e => {
    e.preventDefault();
    store.dispatch({
      type: reduxTypes.aboutModal.ABOUT_MODAL_SHOW
    });
  };

  onLogoutUser = e => {
    const { logoutUser } = this.props;

    e.preventDefault();
    Promise.all([logoutUser()]).then(() => window.location.replace('/'));
  };

  onHelpDropdownToggle = () => {
    const { isHelpDropdownOpen } = this.state;
    this.setState({
      isHelpDropdownOpen: !isHelpDropdownOpen
    });
  };

  onUserDropdownToggle = () => {
    const { isUserDropdownOpen } = this.state;
    this.setState({
      isUserDropdownOpen: !isUserDropdownOpen
    });
  };

  onMobileDropdownToggle = () => {
    const { isMobileDropdownOpen } = this.state;
    this.setState({
      isMobileDropdownOpen: !isMobileDropdownOpen
    });
  };

  onPageResize = size => {
    if (size) {
      this.setState({
        isMobile: size.mobileView
      });
    }
  };

  renderToolbar() {
    const { isUserDropdownOpen, isHelpDropdownOpen } = this.state;
    const { user } = this.props;

    const helpDropdownItems = [
      <DropdownItem key="help0" component="button" onClick={this.onAbout}>
        About
      </DropdownItem>
    ];

    const userDropdownItems = [
      <DropdownItem key="user0" component="button" onClick={this.onLogoutUser}>
        Logout
      </DropdownItem>
    ];

    const helpDropdownToggle = (
      <DropdownToggle onToggle={this.onHelpDropdownToggle} iconComponent={null}>
        <HelpIcon />
      </DropdownToggle>
    );

    const userDropdownToggle = (
      <DropdownToggle onToggle={this.onUserDropdownToggle}>
        <UserIcon /> {user && user.username}
      </DropdownToggle>
    );

    return (
      <Toolbar>
        <ToolbarGroup>
          <ToolbarItem>
            <Dropdown
              isPlain
              position="right"
              onSelect={this.onHelpDropdownToggle}
              isOpen={isHelpDropdownOpen}
              toggle={helpDropdownToggle}
              dropdownItems={helpDropdownItems}
            />
          </ToolbarItem>
          <ToolbarItem>
            <Dropdown
              isPlain
              position="right"
              onSelect={this.onUserDropdownToggle}
              isOpen={isUserDropdownOpen}
              toggle={userDropdownToggle}
              dropdownItems={userDropdownItems}
            />
          </ToolbarItem>
        </ToolbarGroup>
      </Toolbar>
    );
  }

  renderMobileToolbar() {
    const { isMobileDropdownOpen } = this.state;

    const mobileDropdownItems = [
      <DropdownItem key="help0" component="button" onClick={this.onAbout}>
        About
      </DropdownItem>,
      <DropdownItem key="user0" component="button" onClick={this.onLogoutUser}>
        Logout
      </DropdownItem>
    ];

    const mobileDropdownToggle = <KebabToggle onToggle={this.onMobileDropdownToggle} />;

    return (
      <Toolbar>
        <ToolbarGroup>
          <ToolbarItem>
            <Dropdown
              isPlain
              position="right"
              onSelect={this.onMobileDropdownToggle}
              isOpen={isMobileDropdownOpen}
              toggle={mobileDropdownToggle}
              dropdownItems={mobileDropdownItems}
            />
          </ToolbarItem>
        </ToolbarGroup>
      </Toolbar>
    );
  }

  render() {
    const { isMobile } = this.state;
    const { brand, children } = this.props;

    const header = (
      <PageHeader
        logo={<Brand src={brand ? titleImgBrand : titleImg} alt="Cloud Meter Logo" />}
        toolbar={isMobile ? this.renderMobileToolbar() : this.renderToolbar()}
        title="Cloud Meter"
        className="cloudmeter-nav"
      />
    );

    return (
      <React.Fragment>
        <BackgroundImage src={bgImages} />
        <Page className="layout-pf layout-pf-fixed" header={header} onPageResize={this.onPageResize}>
          <PageSection className="cloudmeter-page-section">{children}</PageSection>
        </Page>
      </React.Fragment>
    );
  }
}

PageLayout.propTypes = {
  brand: PropTypes.bool,
  children: PropTypes.node.isRequired,
  logoutUser: PropTypes.func,
  user: PropTypes.shape({
    username: PropTypes.string
  })
};

PageLayout.defaultProps = {
  brand: helpers.RH_BRAND,
  logoutUser: helpers.noop,
  user: {}
};

const mapDispatchToProps = dispatch => ({
  logoutUser: () => dispatch(reduxActions.user.logoutUser())
});

const mapStateToProps = state => ({
  user: state.user.session
});

const ConnectedPageLayout = connect(
  mapStateToProps,
  mapDispatchToProps
)(PageLayout);

export { ConnectedPageLayout as default, ConnectedPageLayout, PageLayout };
