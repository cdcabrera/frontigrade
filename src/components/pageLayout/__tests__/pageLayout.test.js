import React from 'react';
import configureMockStore from 'redux-mock-store';
import { shallow, mount } from 'enzyme';
import { ConnectedPageLayout, PageLayout } from '../pageLayout';

describe('PageLayout Component', () => {
  const generateEmptyStore = (obj = {}) => configureMockStore()(obj);

  it('should render a basic component', () => {
    const store = generateEmptyStore({ user: { session: { username: 'Admin' } } });
    const component = shallow(
      <ConnectedPageLayout>
        <div>test child</div>
      </ConnectedPageLayout>,
      { context: { store } }
    );

    expect(component).toMatchSnapshot('basic');
  });

  it('should render mobile navigation', () => {
    const component = mount(
      <PageLayout>
        <div>test child</div>
      </PageLayout>
    );
    const componentInstance = component.instance();

    component.setState({ isMobile: true });
    expect(component.state().isMobile).toEqual(true);
    expect(component).toMatchSnapshot('mobile collapsed');
    componentInstance.onNavToggle();
    expect(component).toMatchSnapshot('mobile expanded');
  });

  it('should have specific events defined', () => {
    const component = mount(
      <PageLayout>
        <div>test child</div>
      </PageLayout>
    );
    const componentInstance = component.instance();

    expect(componentInstance.onAbout).toBeDefined();
    expect(componentInstance.onHelp).toBeDefined();
    expect(componentInstance.onLogoutUser).toBeDefined();
    expect(componentInstance.onHelpDropdownToggle).toBeDefined();
    expect(componentInstance.onHelpDropdownSelect).toBeDefined();
    expect(componentInstance.onUserDropdownToggle).toBeDefined();
    expect(componentInstance.onUserDropdownSelect).toBeDefined();
    expect(componentInstance.onNavToggle).toBeDefined();
    expect(componentInstance.onPageResize).toBeDefined();
  });

  it('should handle basic events', () => {
    const logoutUser = jest.fn();
    const props = {
      logoutUser
    };
    const component = mount(
      <PageLayout {...props}>
        <div>test child</div>
      </PageLayout>
    );
    const componentInstance = component.instance();
    const mockEvent = { preventDefault: () => {} };

    componentInstance.onLogoutUser(mockEvent);
    expect(logoutUser).toHaveBeenCalled();
  });
});
