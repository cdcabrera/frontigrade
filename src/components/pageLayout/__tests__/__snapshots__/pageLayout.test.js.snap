// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageLayout Component should render a basic component: basic 1`] = `
<PageLayout
  brand={false}
  logoutUser={[Function]}
  user={
    Object {
      "username": "Admin",
    }
  }
>
  <div>
    test child
  </div>
</PageLayout>
`;

exports[`PageLayout Component should render mobile navigation: mobile 1`] = `
<PageLayout
  brand={false}
  logoutUser={[Function]}
  user={Object {}}
>
  <BackgroundImage
    className=""
    src={
      Object {
        "filter": "background-filter.svg#image_overlay",
        "lg": "pfbg_1200.jpg",
        "sm": "pfbg_768.jpg",
        "sm2x": "pfbg_768@2x.jpg",
        "undefined": "pfbg_2000.jpg",
        "xs": "pfbg_576.jpg",
        "xs2x": "pfbg_576@2x.jpg",
      }
    }
  >
    <div
      className="pf-c-background-image css-1p97zyn"
    />
  </BackgroundImage>
  <Page
    className="layout-pf layout-pf-fixed"
    header={
      <PageHeader
        aria-label="Global navigation"
        avatar={null}
        className="cloudmeter-nav"
        isNavOpen={true}
        logo={
          <Brand
            alt="Cloud Meter Logo"
            className=""
            src="title.svg"
          />
        }
        logoProps={null}
        onNavToggle={[Function]}
        showNavToggle={false}
        title="Cloud Meter"
        toolbar={
          <Toolbar
            className={null}
          >
            <ToolbarGroup
              className={null}
            >
              <ToolbarItem
                className={null}
              >
                <Dropdown
                  className=""
                  direction="down"
                  dropdownItems={
                    Array [
                      <Item
                        className=""
                        component="button"
                        href="#"
                        isDisabled={false}
                        isHovered={false}
                        onClick={[Function]}
                      >
                        About
                      </Item>,
                      <Item
                        className=""
                        component="button"
                        href="#"
                        isDisabled={false}
                        isHovered={false}
                        onClick={[Function]}
                      >
                        Logout
                      </Item>,
                    ]
                  }
                  isOpen={false}
                  isPlain={true}
                  onSelect={[Function]}
                  position="right"
                  toggle={
                    <Kebab
                      aria-label="Actions"
                      className=""
                      id=""
                      isActive={false}
                      isDisabled={false}
                      isFocused={false}
                      isHovered={false}
                      isOpen={false}
                      isPlain={false}
                      onToggle={[Function]}
                      parentRef={null}
                    />
                  }
                />
              </ToolbarItem>
            </ToolbarGroup>
          </Toolbar>
        }
        topNav={null}
      />
    }
    isManagedSidebar={false}
    onPageResize={[Function]}
    sidebar={null}
  >
    <div
      className="pf-c-page layout-pf layout-pf-fixed"
    >
      <PageHeader
        aria-label="Global navigation"
        avatar={null}
        className="cloudmeter-nav"
        isNavOpen={true}
        logo={
          <Brand
            alt="Cloud Meter Logo"
            className=""
            src="title.svg"
          />
        }
        logoProps={null}
        onNavToggle={[Function]}
        showNavToggle={false}
        title="Cloud Meter"
        toolbar={
          <Toolbar
            className={null}
          >
            <ToolbarGroup
              className={null}
            >
              <ToolbarItem
                className={null}
              >
                <Dropdown
                  className=""
                  direction="down"
                  dropdownItems={
                    Array [
                      <Item
                        className=""
                        component="button"
                        href="#"
                        isDisabled={false}
                        isHovered={false}
                        onClick={[Function]}
                      >
                        About
                      </Item>,
                      <Item
                        className=""
                        component="button"
                        href="#"
                        isDisabled={false}
                        isHovered={false}
                        onClick={[Function]}
                      >
                        Logout
                      </Item>,
                    ]
                  }
                  isOpen={false}
                  isPlain={true}
                  onSelect={[Function]}
                  position="right"
                  toggle={
                    <Kebab
                      aria-label="Actions"
                      className=""
                      id=""
                      isActive={false}
                      isDisabled={false}
                      isFocused={false}
                      isHovered={false}
                      isOpen={false}
                      isPlain={false}
                      onToggle={[Function]}
                      parentRef={null}
                    />
                  }
                />
              </ToolbarItem>
            </ToolbarGroup>
          </Toolbar>
        }
        topNav={null}
      >
        <header
          className="pf-c-page__header cloudmeter-nav"
          role="banner"
          title="Cloud Meter"
        >
          <div
            className="pf-c-page__header-brand"
          >
            <a
              className="pf-c-page__header-brand-link"
            >
              <Brand
                alt="Cloud Meter Logo"
                className=""
                src="title.svg"
              >
                <img
                  alt="Cloud Meter Logo"
                  className="pf-c-brand"
                  src="title.svg"
                />
              </Brand>
            </a>
          </div>
          <div
            className="pf-c-page__header-tools"
          >
            <Toolbar
              className={null}
            >
              <div
                className="pf-l-toolbar"
              >
                <ToolbarGroup
                  className={null}
                >
                  <div
                    className="pf-l-toolbar__group"
                  >
                    <ToolbarItem
                      className={null}
                    >
                      <div
                        className="pf-l-toolbar__item"
                      >
                        <Dropdown
                          className=""
                          direction="down"
                          dropdownItems={
                            Array [
                              <Item
                                className=""
                                component="button"
                                href="#"
                                isDisabled={false}
                                isHovered={false}
                                onClick={[Function]}
                              >
                                About
                              </Item>,
                              <Item
                                className=""
                                component="button"
                                href="#"
                                isDisabled={false}
                                isHovered={false}
                                onClick={[Function]}
                              >
                                Logout
                              </Item>,
                            ]
                          }
                          isOpen={false}
                          isPlain={true}
                          onSelect={[Function]}
                          position="right"
                          toggle={
                            <Kebab
                              aria-label="Actions"
                              className=""
                              id=""
                              isActive={false}
                              isDisabled={false}
                              isFocused={false}
                              isHovered={false}
                              isOpen={false}
                              isPlain={false}
                              onToggle={[Function]}
                              parentRef={null}
                            />
                          }
                        >
                          <div
                            className="pf-c-dropdown"
                          >
                            <Kebab
                              aria-label="Actions"
                              ariaHasPopup={true}
                              className=""
                              id="pf-toggle-id-4"
                              isActive={false}
                              isDisabled={false}
                              isFocused={false}
                              isHovered={false}
                              isOpen={false}
                              isPlain={true}
                              key=".0"
                              onEnter={[Function]}
                              onToggle={[Function]}
                              parentRef={
                                <div
                                  class="pf-c-dropdown"
                                >
                                  <button
                                    aria-expanded="false"
                                    aria-haspopup="true"
                                    aria-label="Actions"
                                    class="pf-c-dropdown__toggle pf-m-plain"
                                    id="pf-toggle-id-4"
                                    type="button"
                                  >
                                    <svg
                                      aria-hidden="true"
                                      fill="currentColor"
                                      height="1em"
                                      role="img"
                                      style="vertical-align: -0.125em;"
                                      viewBox="0 0 192 512"
                                      width="1em"
                                    >
                                      <path
                                        d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                                        transform=""
                                      />
                                    </svg>
                                  </button>
                                </div>
                              }
                            >
                              <DropdownToggle
                                aria-label="Actions"
                                ariaHasPopup={true}
                                className=""
                                id="pf-toggle-id-4"
                                isActive={false}
                                isDisabled={false}
                                isFocused={false}
                                isHovered={false}
                                isOpen={false}
                                isPlain={true}
                                onEnter={[Function]}
                                onToggle={[Function]}
                                parentRef={
                                  <div
                                    class="pf-c-dropdown"
                                  >
                                    <button
                                      aria-expanded="false"
                                      aria-haspopup="true"
                                      aria-label="Actions"
                                      class="pf-c-dropdown__toggle pf-m-plain"
                                      id="pf-toggle-id-4"
                                      type="button"
                                    >
                                      <svg
                                        aria-hidden="true"
                                        fill="currentColor"
                                        height="1em"
                                        role="img"
                                        style="vertical-align: -0.125em;"
                                        viewBox="0 0 192 512"
                                        width="1em"
                                      >
                                        <path
                                          d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                                          transform=""
                                        />
                                      </svg>
                                    </button>
                                  </div>
                                }
                              >
                                <button
                                  aria-expanded={false}
                                  aria-haspopup={true}
                                  aria-label="Actions"
                                  className="pf-c-dropdown__toggle pf-m-plain"
                                  disabled={false}
                                  id="pf-toggle-id-4"
                                  onClick={[Function]}
                                  onKeyDown={[Function]}
                                  type="button"
                                >
                                  <EllipsisVIcon
                                    color="currentColor"
                                    size="sm"
                                    title={null}
                                  >
                                    <svg
                                      aria-hidden={true}
                                      aria-labelledby={null}
                                      fill="currentColor"
                                      height="1em"
                                      role="img"
                                      style={
                                        Object {
                                          "verticalAlign": "-0.125em",
                                        }
                                      }
                                      viewBox="0 0 192 512"
                                      width="1em"
                                    >
                                      <path
                                        d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                                        transform=""
                                      />
                                    </svg>
                                  </EllipsisVIcon>
                                </button>
                              </DropdownToggle>
                            </Kebab>
                          </div>
                        </Dropdown>
                      </div>
                    </ToolbarItem>
                  </div>
                </ToolbarGroup>
              </div>
            </Toolbar>
          </div>
        </header>
      </PageHeader>
      <main
        className="pf-c-page__main"
        role="main"
      >
        <PageSection
          className="cloudmeter-page-section"
          variant="default"
        >
          <section
            className="pf-c-page__main-section cloudmeter-page-section"
          >
            <div>
              test child
            </div>
          </section>
        </PageSection>
      </main>
    </div>
  </Page>
</PageLayout>
`;
